### ISSUE TITLE
no_std support for futures-io

### ISSUE THREAD

**Issue by Nemo157** [rust-lang-nursery/futures-rs#860]
Inspired by [@Ericson2314's comment](https://github.com/rust-lang-nursery/portability-wg/issues/12#issuecomment-371922047) on a portability-wg issue about allocation-less `std::io::Error` I decided to attempt a two-layered approach to `Async{Read,Write}` allowing supporting `no_std` targets with most of the `futures-io` and `futures-util::io` adaptors.

The main change required is a second set of traits (`CoreAsync{Read,Write}`) which have an associated error type instead of using `std::io::Error`, the `Async{Read,Write}` traits are then implemented on top of these traits where `CoreAsync{Read,Write}::Error: Into<std::io::Error>`.

One difficulty is that some of the more important adaptors (e.g. `futures-util::io::ReadExact`) re-use error variants from `std::io::Error`, my first attempt at working around this limitation is to force the underlying error type to provide at least the two variants that the adaptors require. This seems a bit messy, I'm unsure if there's a much nicer solution to this.

My main goal with this was to be able to use these adaptors with types from the [`embedded-hal`](https://github.com/japaric/embedded-hal) library for non-blocking serial connections, via the [`nb`](https://github.com/japaric/nb) library. I have [an example of this](https://github.com/Nemo157/nrf51-blinker-rs/tree/futures-io) compiling now, but I'll have to wait till I'm at work tomorrow to try and get it running on real hardware.

Links to the changes:
 * [`futures` changes](https://github.com/Nemo157/futures-rs/compare/atomic-std-only...core-io)
 * [`nb` changes](https://github.com/japaric/nb/compare/master...Nemo157:futures-io)
 * [`embedded-hal` changes](https://github.com/japaric/embedded-hal/compare/master...Nemo157:futures-io)

---

I'm creating this issue now to get feedback on whether this is something that might be supported by the core `futures` library at some point. I really would like to avoid splitting the ecosystem between `std` and `no_std` Futures, e.g. I'm likely to be looking at using/building a non-blocking [CoAP](https://en.wikipedia.org/wiki/Constrained_Application_Protocol) library, and am hoping to be able to make it work with both `std` and `no_std` environments without much effort in the library itself.

**Comment by Nemo157** [comment]
And finally got round to using this on some real hardware, it works perfectly (although, integrating task notification into HAL layers without impacting non-`futures` use of them is stil an issue).

**Comment by aturon** [comment]
Thanks @Nemo157 for taking on this work! I *strongly* agree with your overall goals.

I'd really like to see if we can somehow finagle `io::Error` to work in `core` as a general solution to this problem. I will push on that with the libs team.

**Comment by derekdreery** [comment]
It would be really awesome if you could somehow write your embedded code as async functions, and then have some clever executor setup the necessary interrupts and call `wfi` when there's nothing to do. You could possibly end up with code as good (for power usage) as hand-written but with all the embedded details abstracted. Would be very cool.

**Comment by bstrie** [comment]
I'm reviewing issues that are blocking the 0.3 milestone and I'd like a contributor to weigh in on the status of this issue. I'm unclear what action would be needed to resolve this, or whether it's still relevant since its original filing.

**Comment by cramertj** [comment]
I don't believe this is something we'll be able to tackle without support from upstream, since I think it's important that these traits mirror the upstream `std::io::Read/Write` counterparts. I'm open to suggestions here, though, if anyone has ideas about how to avoid this restriction!

**Comment by LiamTheProgrammer** [comment]
@cramertj Not everything can be mirrored for io, but most of it can now thanks to https://github.com/jethrogb/rust-core_io.

**Comment by inthehack** [comment]
We are working on a project  that uses async rust in a no-std environment. I tried to use `embedded-io-async` traits but for now it does not fit ours needs. Their last update is two years old.

What is the status of this issue? Is there need for help?

**Comment by cwahn** [comment]
I do need `AsyncRead` and `AsyncWrite` trait on no_std environment as well. 
What are the main blockers for no_std support?

**Comment by taiki-e** [comment]
There is a PR for unstable `core::io::Error` in upstream: https://github.com/rust-lang/rust/pull/155625

**Comment by protodeniz** [comment]
> There is a PR for unstable `core::io::Error` in upstream: [rust-lang/rust#155625](https://github.com/rust-lang/rust/pull/155625)

i'm not sure if this comment was made before it was merged, or if it matters at all but the pr has been merged now (although still not stable yet)
